### PR TITLE
TDT compute timestamps option and Extra Whitespace handling for SPE

### PR DESCRIPTION
**Pull request by monica-sekoyan**
# What does this PR do ?

This PR includes option for getting timestamps from TDT Decoder head, including on a segment level. This option is enabled for RNNT and CTC decodings as well. And also introduces workaround for getting tokens and id_s for extra whitespaces (even if the tokenizer was not trained in this way.) This is crucial for getting alignments in the future.

**Collection**: asr, common

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* To get timestamps from TDT decoding head, just change the decoding config of the model in the following way:

```python
import nemo.collections.asr.models as nemo_asr
import omegaconf

tdt_asr_model = nemo_asr.EncDecRNNTBPEModel.from_pretrained(model_name='nvidia/parakeet-tdt_ctc-1.1b')

cfg = model.cfg.decoding
with omegaconf.open_dict(cfg):
    cfg['compute_timestamps']=True

tdt_asr_model.change_decoding_strategy(decoding_cfg=cfg)
```

* To control timestamp level manipulate the following field of decoding config:

```python

cfg = model.cfg.decoding
with omegaconf.open_dict(cfg):
    cfg['compute_timestamps']=True
    cfg['rnnt_timestamp_type'] = 'all' # can be "char", "word", "segment" or "all"

tdt_asr_model.change_decoding_strategy(decoding_cfg=cfg)
```

* To make the SPE tokenizer (already trained) encode the text not ignoring occurring extra whitespaces (which is the default behaviour), modify the tokenizer in the following way:

```python
model.tokenizer.ignore_extra_whitespaces = False
```

* To train the SPE tokenizer to make it ignore extra whitespaces both during training and further encoding, use this flag in the training script:

```shell
python scripts/tokenizers/process_asr_text_tokenizer.py --tokenizer=spe --spe_remove_extra_whitespaces
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by nithinraok — APPROVED** [review]
Thanks much!

**Review by nithinraok — DISMISSED** [review]
Minor comments.

File: nemo/collections/asr/parts/submodules/rnnt_decoding.py [hunk 1]
```diff
@@ -867,6 +905,54 @@ def _compute_offsets(
...
+        char_offsets: List[Dict[str, Union[str, int]]],
+        supported_punctuation: Set,
+    ) -> List[Dict[str, Union[str, int]]]:
+
+        ## no refinement for rnnt
+
+        return encoded_char_offsets, char_offsets
+
+    @staticmethod
+    def _refine_timestamps_tdt(
```
[the remark this answers is not in the document]
Comment (monica-sekoyan) [reply nemo/collections/asr/parts/submodules/rnnt_decoding.py:938]: we use it in `_get_segment_offsets` (maybe only partially, but I guess in the future it can be extended). if we want to compute `supported_punctuation` here, we need to pass `vocabulary` to decoding object anyway(which will also only be used for this case). any difference to pass `vocabulary` or `supported_punctuation` through decoding object ?

**Review by nithinraok — DISMISSED** [review]

File: nemo/collections/asr/parts/submodules/ctc_decoding.py [hunk 2]
```diff
@@ -858,6 +902,92 @@ def _get_word_offsets_subwords_sentencepiece(
 
         return word_offsets
 
+    @staticmethod
+    def _get_segment_offsets(
+        offsets: Dict[str, Union[str, float]],
+        segment_delimiter_tokens: List[str],
+        supported_punctuation: Set = None,
+        segment_gap_threshold: int = None,
```
Comment (nithinraok) [inline nemo/collections/asr/parts/submodules/ctc_decoding.py:910]: int=None?

File: nemo/collections/asr/parts/submodules/ctc_decoding.py [hunk 3]
```diff
@@ -858,6 +902,92 @@ def _get_word_offsets_subwords_sentencepiece(
...
+        Returns:
+            A list of dictionaries containing the segment offsets. Each item contains "segment", "start_offset" and
+            "end_offset".
+        """
+        if (
+            supported_punctuation
+            and not set(segment_delimiter_tokens).intersection(supported_punctuation)
+            and not segment_gap_threshold
+        ):
+            logging.warning(
```
Comment (nithinraok) [inline nemo/collections/asr/parts/submodules/ctc_decoding.py:930]: Why segment_gap_threshold condition is checked here? 
Log this only once using logging MODE=ONCE, otherwise we get this logging for each utterance.

File: nemo/collections/asr/parts/submodules/ctc_decoding.py (hunk 2 above)
Comment (monica-sekoyan) [reply nemo/collections/asr/parts/submodules/ctc_decoding.py]: corrected all the occurencies

File: tests/collections/common/prompt_formatters/test_canary_prompt_formatter.py [hunk 4]
```diff
@@ -20,9 +20,9 @@ def test_canary_prompt_formatter_training(canary_tokenizer):
     )
     assert set(ans) == {"input_ids", "context_ids", "answer_ids", "mask"}
     # fmt: off
-    assert ans["input_ids"].tolist() == [4, 8, 7, 8, 5, 11, 91, 30, 40, 3]
-    assert ans["context_ids"].tolist() == [4, 8, 7, 8, 5]
-    assert ans["answer_ids"].tolist() == [11, 91, 30, 40, 3]
+    assert canary_tokenizer.ids_to_text(ans["input_ids"].tolist()) == '<|startoftranscript|><|en|><|transcribe|><|en|><|pnc|> TEST<|endoftext|>'
```
Comment (pzelasko) [inline tests/collections/common/prompt_formatters/test_canary_prompt_formatter.py:23]: Thank you, these prompt formatting tests are much better now - readable and future-proof.

File: nemo/collections/common/tokenizers/sentencepiece_tokenizer.py [hunk 5]
```diff
@@ -36,13 +37,19 @@ class SentencePieceTokenizer(TokenizerSpec, ChatTemplateMixin):
         special_tokens: either list of special tokens or dictionary of token name to token value
         legacy: when set to True, the previous behavior of the SentecePiece wrapper will be restored,
             including the possibility to add special tokens inside wrapper.
+        ignore_extra_whitespaces: whether to ignore extra whitespaces in the input text while encoding.
```
Comment (pzelasko) [inline nemo/collections/common/tokenizers/sentencepiece_tokenizer.py:40]: I'm not sure I understand - why is this needed?

Comment (monica-sekoyan) [reply nemo/collections/common/tokenizers/sentencepiece_tokenizer.py:40]: This is for already trained tokenizers (which are trained to ignore extra whitespaces). by adding this parameter we can make inference to these models not to ignore them.
So, for example, if we want to encode ASR model's prediction `text with     a lot of spaces`, we will get this tokens (and their corresponding ids) `["_text", "_with", "_", "a", "_lot", "_of", "_spaces"]`.  but instead we should have got  `["_text", "_with", "_",  "_",  "_",  "_", "a", "_lot", "_of", "_spaces"]` which was in fact predicted by the model and can be used for alignment purposes (I believe there are other use-cases too)

**Review by nithinraok — APPROVED** [review]
LGTM, Thank you!
